NullPointerException in AbstractRouteRegistryInitializer.validateRouteAnnotation on WebSphere 8.5.5.14

We are migrating our application from Vaadin 8 to Vaadin 12. Unfortunately the application refuses to deploy on IBM WebSphere 8.5.5.14.

The following NullPointerException occurs during application deployment.

`com.ibm.ws.exception.RuntimeWarning: com.ibm.ws.exception.RuntimeWarning: com.ibm.ws.webcontainer.exception.WebAppNotLoadedException: Failed to load webapp: Failed to load webapp: null
	at com.ibm.wsspi.runtime.component.WsComponentImpl.join(WsComponentImpl.java:432)
	at com.ibm.ws.runtime.component.CompositionUnitMgrImpl.startCUList(CompositionUnitMgrImpl.java:1295)
	at com.ibm.ws.runtime.component.CompositionUnitMgrImpl.start(CompositionUnitMgrImpl.java:275)
	at com.ibm.ws.runtime.component.ContainerHelper.startComponents(ContainerHelper.java:540)
	at com.ibm.ws.runtime.component.ContainerImpl.startComponents(ContainerImpl.java:627)
	at com.ibm.ws.runtime.component.ContainerImpl.start(ContainerImpl.java:618)
	at com.ibm.ws.runtime.component.ApplicationServerImpl.start(ApplicationServerImpl.java:252)
	at com.ibm.ws.runtime.component.ContainerHelper.startComponents(ContainerHelper.java:540)
	at com.ibm.ws.runtime.component.ContainerImpl.startComponents(ContainerImpl.java:627)
	at com.ibm.ws.runtime.component.ContainerImpl.start(ContainerImpl.java:618)
	at com.ibm.ws.runtime.component.ServerImpl.start(ServerImpl.java:540)
	at com.ibm.ws.runtime.WsServerImpl.bootServerContainer(WsServerImpl.java:319)
	at com.ibm.ws.runtime.WsServerImpl.start(WsServerImpl.java:232)
	at com.ibm.ws.runtime.WsServerImpl.main(WsServerImpl.java:705)
	at com.ibm.ws.runtime.WsServer.main(WsServer.java:59)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at com.ibm.wsspi.bootstrap.WSLauncher.launchMain(WSLauncher.java:234)
	at com.ibm.wsspi.bootstrap.WSLauncher.main(WSLauncher.java:96)
	at com.ibm.wsspi.bootstrap.WSLauncher.run(WSLauncher.java:77)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.eclipse.equinox.internal.app.EclipseAppContainer.callMethodWithException(EclipseAppContainer.java:587)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:198)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:110)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:79)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:369)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:179)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.eclipse.core.launcher.Main.invokeFramework(Main.java:340)
	at org.eclipse.core.launcher.Main.basicRun(Main.java:282)
	at org.eclipse.core.launcher.Main.run(Main.java:981)
	at com.ibm.wsspi.bootstrap.WSPreLauncher.launchEclipse(WSPreLauncher.java:411)
	at com.ibm.wsspi.bootstrap.WSPreLauncher.main(WSPreLauncher.java:173)
Caused by: com.ibm.ws.exception.RuntimeWarning: com.ibm.ws.webcontainer.exception.WebAppNotLoadedException: Failed to load webapp: Failed to load webapp: null
	at com.ibm.ws.webcontainer.component.WebContainerImpl.install(WebContainerImpl.java:433)
	at com.ibm.ws.webcontainer.component.WebContainerImpl.start(WebContainerImpl.java:719)
	at com.ibm.ws.runtime.component.ApplicationMgrImpl.start(ApplicationMgrImpl.java:1211)
	at com.ibm.ws.runtime.component.DeployedApplicationImpl.fireDeployedObjectStart(DeployedApplicationImpl.java:1450)
	at com.ibm.ws.runtime.component.DeployedModuleImpl.start(DeployedModuleImpl.java:639)
	at com.ibm.ws.runtime.component.DeployedApplicationImpl.start(DeployedApplicationImpl.java:1032)
	at com.ibm.ws.runtime.component.ApplicationMgrImpl.startApplication(ApplicationMgrImpl.java:795)
	at com.ibm.ws.runtime.component.ApplicationMgrImpl$5.run(ApplicationMgrImpl.java:2279)
	at com.ibm.ws.security.auth.ContextManagerImpl.runAs(ContextManagerImpl.java:5482)
	at com.ibm.ws.security.auth.ContextManagerImpl.runAsSystem(ContextManagerImpl.java:5698)
	at com.ibm.ws.security.core.SecurityContext.runAsSystem(SecurityContext.java:255)
	at com.ibm.ws.runtime.component.ApplicationMgrImpl.start(ApplicationMgrImpl.java:2284)
	at com.ibm.ws.runtime.component.CompositionUnitMgrImpl.start(CompositionUnitMgrImpl.java:436)
	at com.ibm.ws.runtime.component.CompositionUnitImpl.start(CompositionUnitImpl.java:123)
	at com.ibm.ws.runtime.component.CompositionUnitMgrImpl.start(CompositionUnitMgrImpl.java:379)
	at com.ibm.ws.runtime.component.CompositionUnitMgrImpl.access$500(CompositionUnitMgrImpl.java:127)
	at com.ibm.ws.runtime.component.CompositionUnitMgrImpl$CUInitializer.run(CompositionUnitMgrImpl.java:985)
	at com.ibm.wsspi.runtime.component.WsComponentImpl$_AsynchInitializer.run(WsComponentImpl.java:524)
	at com.ibm.ws.util.ThreadPool$Worker.run(ThreadPool.java:1892)
Caused by: com.ibm.ws.webcontainer.exception.WebAppNotLoadedException: Failed to load webapp: Failed to load webapp: null
	at com.ibm.ws.webcontainer.WSWebContainer.addWebApp(WSWebContainer.java:914)
	at com.ibm.ws.webcontainer.WSWebContainer.addWebApplication(WSWebContainer.java:789)
	at com.ibm.ws.webcontainer.component.WebContainerImpl.install(WebContainerImpl.java:427)
	... 18 more
Caused by: com.ibm.ws.webcontainer.exception.WebAppNotLoadedException: Failed to load webapp: null
	at com.ibm.ws.webcontainer.VirtualHostImpl.addWebApplication(VirtualHostImpl.java:186)
	at com.ibm.ws.webcontainer.WSWebContainer.addWebApp(WSWebContainer.java:904)
	... 20 more
Caused by: java.lang.NullPointerException
	at com.vaadin.flow.server.startup.AbstractRouteRegistryInitializer.validateRouteAnnotation(AbstractRouteRegistryInitializer.java:191)
	at com.vaadin.flow.server.startup.AbstractRouteRegistryInitializer.lambda$checkForConflictingAnnotations$1(AbstractRouteRegistryInitializer.java:97)
	at com.vaadin.flow.server.startup.AbstractRouteRegistryInitializer$$Lambda$31.000000001F8D7EA0.accept(Unknown Source)
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:959)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:591)
	at com.vaadin.flow.server.startup.AbstractRouteRegistryInitializer.checkForConflictingAnnotations(AbstractRouteRegistryInitializer.java:93)
	at com.vaadin.flow.server.startup.AbstractRouteRegistryInitializer$$Lambda$24.0000000011274F90.accept(Unknown Source)
	at java.util.stream.ReferencePipeline$11$1.accept(ReferencePipeline.java:383)
	at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1564)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:523)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:513)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:719)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:245)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:510)
	at com.vaadin.flow.server.startup.AbstractRouteRegistryInitializer.validateRouteClasses(AbstractRouteRegistryInitializer.java:63)
	at com.vaadin.flow.server.startup.RouteRegistryInitializer.onStartup(RouteRegistryInitializer.java:47)
	at com.ibm.ws.webcontainer.webapp.WebAppImpl.initializeServletContainerInitializers(WebAppImpl.java:620)
	at com.ibm.ws.webcontainer.webapp.WebAppImpl.initialize(WebAppImpl.java:410)
	at com.ibm.ws.webcontainer.webapp.WebGroupImpl.addWebApplication(WebGroupImpl.java:88)
	at com.ibm.ws.webcontainer.VirtualHostImpl.addWebApplication(VirtualHostImpl.java:171)
	... 21 more`

The NullPointerException occurs because the Route interface itself comes as the parameter route in the AbstractRouteRegistryInitializer.validateRouteAnnotation method.  This occurs when trying to deploy our application on WebSphere.  It works fine on Glassfish.

I have investigated the problem and my conclusion is that WebSphere handled the @HandlesTypes({ Route.class, RouteAlias.class }) in the RouteRegistryInitializer class differently than Glassfish.  The variable classSet includes the class for both the Route and RouteAlias interface on Websphere but on Glassfish it is only classes that uses the Route annotation that are included.

I believe that the problem can be solved by filtering out the Route and RouterInterface in the AbstractRouteRegistryInitializer.validateRouteAnnotation method.

Vaadin version: 12.0.1
WebSphere version: 8.5.5.14
Java version: 1.8_64
OS: Windows 10

--- a/flow-server/src/main/java/com/vaadin/flow/server/startup/AbstractRouteRegistryInitializer.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/startup/AbstractRouteRegistryInitializer.java
@@ -132,7 +132,7 @@ public abstract class AbstractRouteRegistryInitializer implements Serializable {
     private void validateRouteImplementation(Class<?> route,
             Class<?> implementation) {
         Route annotation = route.getAnnotation(Route.class);
-        if (!UI.class.equals(annotation.layout())) {
+        if (annotation != null && !UI.class.equals(annotation.layout())) {
             if (implementation.isAssignableFrom(route)) {
                 throw new InvalidRouteLayoutConfigurationException(String
                         .format("%s needs to be the top parent layout '%s' not '%s'",
@@ -206,7 +206,8 @@ public abstract class AbstractRouteRegistryInitializer implements Serializable {
     private void validateRouteAnnotation(Class<?> route,
             Class<? extends Annotation> annotation) {
         Route routeAnnotation = route.getAnnotation(Route.class);
-        if (!UI.class.equals(routeAnnotation.layout())) {
+        if (routeAnnotation != null
+                && !UI.class.equals(routeAnnotation.layout())) {
             if (route.isAnnotationPresent(annotation)) {
                 throw new InvalidRouteLayoutConfigurationException(String
                         .format("%s annotation needs to be on the top parent layout '%s' not on '%s'",

